src/slice-assign.c

```diff
@@ -378,25 +379,27 @@ static SEXP vec_assign_fallback(SEXP x, SEXP index, SEXP value) {
 }
 
 static
-SEXP vec_proxy_assign_names(SEXP proxy, SEXP index, SEXP value) {
+SEXP vec_proxy_assign_names(SEXP proxy,
+                            SEXP index,
+                            SEXP value,
+                            const enum vctrs_owned owned) {
   SEXP value_nms = PROTECT(vec_names(value));
 
   if (value_nms == R_NilValue) {
     UNPROTECT(1);
     return proxy;
   }
 
-  SEXP proxy_nms = PROTECT(vec_names(proxy));
+  SEXP proxy_nms = PROTECT(vec_proxy_names(proxy));
   if (proxy_nms == R_NilValue) {
     proxy_nms = PROTECT(Rf_allocVector(STRSXP, vec_size(proxy)));
   } else {
-    proxy_nms = PROTECT(r_clone_referenced(proxy_nms));
+    proxy_nms = PROTECT(vec_clone_referenced(proxy_nms, owned));
   }
+  proxy_nms = PROTECT(chr_assign(proxy_nms, index, value_nms, owned));
 
-  proxy_nms = PROTECT(chr_assign(proxy_nms, index, value_nms, VCTRS_OWNED_true));
```

Comment:
I always found it a bit odd that `VCTRS_OWNED_true` was constant here, so this is a nice change 👍